common : use common_chat_templates for add_bos and add_eos

danbev opened the pull request:
This commit updates common_chat_templates_apply_jinja to use the
the add_bos and add_eos parameters from the chat template instead of
the inputs.
    
The motivation for this is that currently if the `add_bos` and `add_eos`
from the input parameters are used it is possible to there will be a
missmatch between the model and the chat template which can lead to the
the removal of duplicate BOS/EOS tokens in chat.cpp `apply` to not
happen leading to two BOS tokens being added to the template.

------
I've tried this using new converted models and the bos duplication is not there. If this solution is accepted then I'll re-convert the instruction tuned models and upload them to [ggml-org](https://huggingface.co/collections/ggml-org/gemma-3-270m-689e0105d56462786413d7fc).

CISC commented:
You're saying double BOS is being added to the instruction tuned model, but only without jinja?

I can't verify the model config as it's gated, though looking at f.ex. MLX versions it seems BOS is `<bos>`

So that doesn't make much sense, the `gemma` chat template doesn't have BOS, and SPM has `add_bos` set by default, meaning there should be only one BOS being added. For jinja there's a BOS in the chat template, but as long as `add_bos` is `true` this should be automatically removed.

ggerganov commented:
Here is my understanding:

- Without this change, if you run IT model without `--jinja` it uses single BOS (correct). But if you add `--jinja` it will have 2 BOS (wrong)
- With this change, if you run the IT model without `--jinja` it uses zero BOS (wrong) and with `--jinja` it uses 1 BOS (correct)

Maybe I don't understand the logic completely, but this seems very confusing. I can't tell when `--jinja` should be used and when it should not. Can we improve this somehow?

danbev commented:
Sorry about the confusion, it was late yesterday and I was a little rushed creating this PR. I've not looked at this part of the code base much, but I'll take a closer look today and try to understand this issue better.

danbev commented:
I noticed that the instruction tuned model has the following:
```console
(venv) $ head ~/work/ai/models/gemma-3-270m-it/tokenizer_config.json
{
  "add_bos_token": true,
  "add_eos_token": false,
  "added_tokens_decoder": {
   ...
```
The pretrained/base model also `add_bos_token` set to true which I think is correct, but I don't think this should be `true` for the instruction tuned model?

CISC commented:
@danbev Yep, you are right, I overlooked this codepath.

danbev commented:
> Edit: Sorry for the back-and-forth, but I think this is the only change that needs to be done, the cause of the problem is here:

No worries at all! Sounds good, I'll try that, thanks!

danbev committed:
common : use common_chat_templates for add_bos and add_eos

This commit updates common_chat_templates_apply_jinja to use the
the add_bos and add_eos parameters from the chat template instead of
the inputs.

The motivation for this is that currently if the `add_bos` and `add_eos`
from the input parameters are used it is possible to there will be a
missmatch between the model and the chat template which can lead to the
the removal of duplicate BOS/EOS tokens in chat.cpp `apply` to not
happen leading to two BOS tokens being added to the template.

CISC reviewed (approved):
Thanks, everything works now with/without jinja?

broadbit-hu commented:
Already tested with b6152 relase and Mistral NeMo Instruct:
```
./build/bin/llama-cli -m ../models/mistral-nemo-instruct-2407-q8_0.gguf -c 0 -fa --jinja -p "Test" --verbose-prompt
```

Results:
```
check_double_bos_eos: Added a BOS token to the prompt as specified by the model but the prompt also starts with a BOS token. So now the final prompt starts with 2 BOS tokens. Are you sure this is what you want?
main: prompt: 'Test'
main: number of tokens in prompt = 5
     1 -> '<s>'
     1 -> '<s>'
     3 -> '[INST]'
  4922 -> 'Test'
     4 -> '[/INST]'
```

broadbit-hu commented:
> @broadbit-hu Could you give this a try with [b6178](https://github.com/ggml-org/llama.cpp/releases/tag/b6178)? This is the release that contains this code of this PR.

It's perfect (tested with Mistral NeMo), thanks for the fix! :)

```
main: number of tokens in prompt = 4
     1 -> '<s>'
     3 -> '[INST]'
  4922 -> 'Test'
     4 -> '[/INST]'
```

Something is wrong with Mistral Small yet:
```
./build/bin/llama-cli -m ../models/mistral-small-24b-instruct-2506-q4_k_m.gguf -c 0 -fa --jinja -p "Test" --verbose-prompt
```
```
print_info: model params     = 23.57 B
print_info: general.name     = Mistral Small 3.2 24B Instruct 2506
print_info: vocab type       = BPE
print_info: n_vocab          = 131072
print_info: n_merges         = 269443
print_info: BOS token        = 1 '<s>'
print_info: EOS token        = 2 '</s>'
print_info: UNK token        = 0 '<unk>'
print_info: PAD token        = 11 '<pad>'
print_info: LF token         = 1010 'Ċ'
print_info: EOG token        = 2 '</s>'

...

Failed to infer a tool call example (possible template bug)
main: llama threadpool init, n_threads = 6
main: chat template is available, enabling conversation mode (disable it with -no-cnv)
*** User-specified prompt will pre-start conversation, did you mean to set --system-prompt (-sys) instead?
main: chat template example:
mistral-v7-tekken

main: prompt: 'Test'
main: number of tokens in prompt = 8
     1 -> '<s>'
 39575 -> 'mist'
  2784 -> 'ral'
  9332 -> '-v'
  1055 -> '7'
  1045 -> '-'
 17848 -> 'tek'
  3569 -> 'ken'
```

CISC commented:
> Something is wrong with Mistral Small yet:

Yeah, the template fix-hack backfires when used with `--jinja`, someone should take a look at that...

danbev commented:
> Yeah, the tekken template fix-hack backfires when used with `--jinja`, someone should take a look at that...

I'll take a closer look at this next week :+1:

danbev commented:
@broadbit-hu Great, glad to hear that! I also tried out the unsloth model yesterday and it worked, which was the reason for asking about the actual model you were using.
